### PR TITLE
Patatrack: introduce pt-clamping in vertex sorting, add #tracks in track SOA

### PR DESCRIPTION
**Pull request by VinInn**
This PR introduces some long standing improvements in Patatrack that can be useful to POG and PAG using directly SOA products.

1) Track Pt-clamping is introduced in vertex sorting.
  This is supposed to mitigate the effect of spurious tracks with large pt due to miss-measurement, resolution or inaccurate calibration/alignment.

The effect has been tested in Z->tau-tau events that are supposed to be among those who most suffer from signal-vertex miss-identification.
The effect is small if not negligible. see: http://innocent.home.cern.ch/innocent/RelVal/ztt_clamp/plots_selectedPixelVertices/pvtagging.pdf
where we compare clamping at 20, 75 (the default) and 1000 GeV.
This is not unexpected given the resolution of HP quadruplets, those used to build vertices, that reaches ~30% at ~70 GeV.
It may be different with initial calibrations or large inefficiency.

Results for ttbar can be found in
http://innocent.home.cern.ch/innocent/RelVal/gpuMTVclamp/plots_selectedPixelVertices/pvtagging.pdf

2) The number of tracks is stored in the track SOA.
This should make the use of the SOA more user-friendly.

Default has changed, so expect small regressions.

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild, please test

**Comment by VinInn** [comment]
enable gpu

**Comment by VinInn** [comment]
what is waiting for?

**Comment by smuzaffar** [comment]
@mmusich , jenkins was very bus today that is why there was not enough resources to run tests, I see that https://cmssdt.cern.ch/jenkins/job/compare-root-files-short-matrix/49033/console is still running for last 2 hours. Hopefully it will finish soon.

**Comment by jpata** [comment]
assign @cms-sw/tracking-pog-l2

**Comment by mmusich** [comment]
assign tracking-pog

assuming that was the intention of https://github.com/cms-sw/cmssw/pull/37281#issuecomment-1075296145

**Comment by mmusich** [comment]
+1
   * @cms-sw/reconstruction-l2 what is being waited upon here?

**Comment by jpata** [comment]
Thanks Marco, I wanted to ask tracking to take a look (and make sure it was discussed in the POG) as I didn't really have any comments. I will sign now.

**Comment by jpata** [comment]
+reconstruction

File: RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorKernelsImpl.h [hunk 1]
```diff
@@ -557,13 +557,15 @@ __global__ void kernel_fillHitDetIndices(HitContainer const *__restrict__ tuples
...
 
-__global__ void kernel_fillNLayers(TkSoA *__restrict__ ptracks) {
+__global__ void kernel_fillNLayers(TkSoA *__restrict__ ptracks, cms::cuda::AtomicPairCounter *apc) {
   auto &tracks = *ptracks;
   auto first = blockIdx.x * blockDim.x + threadIdx.x;
-  for (int idx = first, nt = TkSoA::stride(); idx < nt; idx += gridDim.x * blockDim.x) {
+  auto ntracks = apc->get().m;
+  if (0 == first)
+    tracks.setNTracks(ntracks);
+  for (int idx = first, nt = ntracks; idx < nt; idx += gridDim.x * blockDim.x) {
```
Comment (fwyzard) [inline RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorKernelsImpl.h:566]: just checking -- why is `nt` needed ? wouldn't `idx < ntracks` equivalent ?

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by missirol** [comment]
> Default has changed, so expect small regressions.

@VinInn @fwyzard @silviodonato , should this PR be backported to `12_3_X`? (since HLT will use that release at the start of Run 3)

**Comment by silviodonato** [comment]
I definitely support the request of a backport, if it is possible. It would be very good if we can start the data taking including this PR.

**Comment by perrotta** [comment]
+1

**Pull request merged**